bundle/src/main/java/com/adobe/acs/commons/fetchfile/impl/FileFetchImpl.java

```diff
@@ -0,0 +1,229 @@
+      }
+    } finally {
+      if (con != null) {
+        con.disconnect();
+      }
+    }
+  }
+
+  @Override
+  public Exception getLastException() {
+    return this.lastException;
+  }
+
+  private String hash(byte[] data) {
+    return new String(Base64.encode(DigestUtils.getMd5Digest().digest(data)));
+  }
+
+  @Override
+  public boolean isLastJobSucceeded() {
+    return this.lastJobSucceeded;
+  }
+
+  protected HttpURLConnection openConnection() throws IOException {
+    return (HttpURLConnection) new URL(config.remoteUrl()).openConnection();
+  }
+
+  @Override
+  public void run() {
+    try {
+      updateFile();
+      lastJobSucceeded = true;
+      lastException = null;
+    } catch (Exception e) {
+      log.warn("Failed to run fetch file job for {}", config.remoteUrl(), e);
+      lastException = e;
+      lastJobSucceeded = false;
+    }
+  }
+
+  public void setFactory(ResourceResolverFactory factory) {
+    this.factory = factory;
+  }
+
+  public void setReplicator(Replicator replicator) {
+    this.replicator = replicator;
+  }
+
+  @Override
+  public void updateFile() throws IOException, ReplicationException {
+    log.trace("updateFile");
+
+    try (ResourceResolver resolver = factory
+        .getServiceResourceResolver(Collections.singletonMap(ResourceResolverFactory.SUBSERVICE, SERVICE_USER_NAME))) {
+      AssetManager manager = Optional.ofNullable(resolver.adaptTo(AssetManager.class))
+          .orElseThrow(() -> new PersistenceException("Failed to get Asset Manager"));
+
+      byte[] bin = fetchFile();
+
+      boolean publish = false;
+      String path = config.damPath();
+      Resource assetResource = resolver.getResource(path);
+
+      String fileHash = hash(bin);
+      Asset asset = null;
+      if (assetResource != null) {
+        asset = Optional.ofNullable(assetResource.adaptTo(Asset.class))
+            .orElseThrow(() -> new PersistenceException("Failed to adapt Resource to Asset"));
+
+        String assetHash = assetResource.getValueMap().get("jcr:content/metadata/hash", "");
+        if (!fileHash.equals(assetHash)) {
+          log.info("File does not match, updating...");
+          asset.removeRendition("original");
+          asset.addRendition("original", new ByteArrayInputStream(bin), Collections.emptyMap());
+          updateHash(fileHash, assetResource);
+          publish = true;
+        } else {
+          log.info("Hashes match, not updating");
+        }
+      } else {
+        log.info("Adding new file...");
+        asset = manager.createAsset(path, new ByteArrayInputStream(bin), config.mimeType(), true);
+        assetResource = Optional.ofNullable(asset.adaptTo(Resource.class))
+            .orElseThrow(() -> new PersistenceException("Failed to adapt Asset to Resource"));
+        updateHash(fileHash, assetResource);
+        publish = true;
+      }
+
+      if (publish) {
+        log.info("Replicating...");
```

Comment:
can you add the path of the asset to the log message?

Reply:
Added here and generally cleaned up the log messages